File: lab2pbudhath/app/src/main/java/com/example/lab2_pbudhath/MainActivity.java
package com.example.lab2_pbudhath;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {
    private double bill = 0.0;
    private double service_quality = 0.0;
    private double exc_quality = 0.25, avg_quality = 0.2, bavg_quality = 0.15;
    private int num_people = 1;
    private boolean radio_checked = false;

    public final static String RESULT1 = "com.example.lab2_pbudhath.R1";
    public final static String RESULT2 = "com.example.lab2_pbudhath.R2";
    public final static String RESULT3 = "com.example.lab2_pbudhath.R3";

    public final static String STATE_EXC_QUAL = "exc_quality";
    public final static String STATE_AVG_QUAL = "avg_quality";
    public final static String STATE_BAVG_QUAL = "bavg_quality";
    public final static String STATE_RADIO_CHECK = "radio_checked";
    public final static String STATE_NUM_PEOPLE = "num_people";
    public final static String STATE_SERVICE_QUAL = "service_quality";
    public final static String STATE_BILL = "bill";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if (savedInstanceState != null){
            exc_quality = savedInstanceState.getDouble(STATE_EXC_QUAL);
            avg_quality = savedInstanceState.getDouble(STATE_AVG_QUAL);
            bavg_quality = savedInstanceState.getDouble(STATE_BAVG_QUAL);

            service_quality = savedInstanceState.getDouble(STATE_SERVICE_QUAL);
            bill = savedInstanceState.getDouble(STATE_BILL);
            num_people = savedInstanceState.getInt(STATE_NUM_PEOPLE);
            radio_checked = savedInstanceState.getBoolean(STATE_RADIO_CHECK);
            
            updateServiceQuality();
            updateText();
        }
        EditText Bill_Edit = findViewById(R.id.Bill_Edit);
        EditText People_Edit = findViewById(R.id.People_Edit);
        TextWatcher twB = new TextWatcher(){
            public void afterTextChanged(Editable s) {}
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (Bill_Edit.getText().length() != 0) {
                    bill = Float.parseFloat(String.valueOf(Bill_Edit.getText()));
                    if(bill > 10000){
                        Toast.makeText(getApplicationContext(), "Max is $10,000", Toast.LENGTH_SHORT).show();
                        bill = 0;
                        return;
                    }
                    if(bill < 0){
                        Toast.makeText(getApplicationContext(), "Invalid Entry", Toast.LENGTH_SHORT).show();
                        bill = 0;
                        return;
                    }
                    if(radio_checked){
                        updateText();
                    }
                    else{
                        Toast.makeText(getApplicationContext(), "Select a service quality", Toast.LENGTH_SHORT).show();
                    }
                }
                else {
                    bill = 0.0;
                    ((TextView)findViewById(R.id.Tip_Total_Val)).setText("");
                    ((TextView)findViewById(R.id.Total_Val)).setText("");
                    ((TextView)findViewById(R.id.Total_Per_Val)).setText("");
                }
            }
        };
        Bill_Edit.addTextChangedListener(twB);
        Bill_Edit.setText("");

        TextWatcher twP = new TextWatcher(){
            public void afterTextChanged(Editable s) {}
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (People_Edit.getText().length() != 0) {
                    num_people = Integer.parseInt(String.valueOf(People_Edit.getText()));
                    if(num_people > 10){
                        Toast.makeText(getApplicationContext(), "Max 10 people", Toast.LENGTH_SHORT).show();
                        num_people = 1;
                        return;
                    }
                    if(num_people < 1){
                        Toast.makeText(getApplicationContext(), "Invalid Entry", Toast.LENGTH_SHORT).show();
                        num_people = 1;
                        return;
                    }
                    if(radio_checked && bill != 0){
                        updateText();
                    }
                    else{
                        Toast.makeText(getApplicationContext(), "Select a service quality", Toast.LENGTH_SHORT).show();
                    }
                }
                else {
                    num_people = 1;
                    ((TextView)findViewById(R.id.Tip_Total_Val)).setText("");
                    ((TextView)findViewById(R.id.Total_Val)).setText("");
                    ((TextView)findViewById(R.id.Total_Per_Val)).setText("");
                }
            }
        };
        People_Edit.addTextChangedListener(twP);
        People_Edit.setText("");

    }

    public void myOnClick(View v){
        if(v.getId() == R.id.Clear_Button){
            ((EditText)findViewById(R.id.Bill_Edit)).setText("");
            ((EditText)findViewById(R.id.People_Edit)).setText("");
            ((RadioButton)findViewById(R.id.Radio_Exc)).setChecked(false);
            ((RadioButton)findViewById(R.id.Radio_Avg)).setChecked(false);
            ((RadioButton)findViewById(R.id.Radio_Bavg)).setChecked(false);
            ((TextView)findViewById(R.id.Tip_Total_Val)).setText("");
            ((TextView)findViewById(R.id.Total_Val)).setText("");
            ((TextView)findViewById(R.id.Total_Per_Val)).setText("");
            radio_checked = false;
            bill = 0.0;
        }

        if(v.getId() == R.id.Update_Button){
            Intent intent = new Intent(this, ServiceSelectActivity.class);
            intent.putExtra(ServiceSelectActivity.PARAM1, exc_quality);
            intent.putExtra(ServiceSelectActivity.PARAM2, avg_quality);
            intent.putExtra(ServiceSelectActivity.PARAM3, bavg_quality);
            someActivityResultLauncher.launch(intent);

        }
    }
    public void onServiceRadioClick(View v){
        boolean check = ((RadioButton)v).isChecked();
        radio_checked = check;
        if(!check){
            service_quality = 0.0;
            return;
        }
        if(v.getId() == R.id.Radio_Exc){
            service_quality = exc_quality;
        }

        if(v.getId() == R.id.Radio_Avg){
            service_quality = avg_quality;
        }

        if(v.getId() == R.id.Radio_Bavg){
            service_quality = bavg_quality;
        }

        if(bill == 0){
            Toast.makeText(getApplicationContext(), "Enter bill amount", Toast.LENGTH_SHORT).show();
        }
        else{
            updateText();
        }
    }

    private void updateText(){
        
        if(!radio_checked){
            return;
        }
        
        if(bill == 0.0){
            return;
        }

        double tip_total =  bill * service_quality;
        double total = bill + tip_total;
        double total_per = total/num_people;

        ((TextView)findViewById(R.id.Tip_Total_Val)).setText(String.format("%.2f%n", tip_total));
        ((TextView)findViewById(R.id.Total_Val)).setText(String.format("%.2f%n", total));
        ((TextView)findViewById(R.id.Total_Per_Val)).setText(String.format("%.2f%n", total_per));
    }

    private void updateServiceQuality(){
        ((TextView)findViewById(R.id.Radio_Exc)).setText(String.format("Excellent (%.2f%%)", exc_quality*100));
        ((TextView)findViewById(R.id.Radio_Avg)).setText(String.format("Average (%.2f%%)", avg_quality*100));
        ((TextView)findViewById(R.id.Radio_Bavg)).setText(String.format("Below Average (%.2f%%)", bavg_quality*100));
    }

    @Override
    public void onSaveInstanceState(Bundle savedInstanceState) {
        savedInstanceState.putDouble(STATE_EXC_QUAL, exc_quality);
        savedInstanceState.putDouble(STATE_AVG_QUAL, avg_quality);
        savedInstanceState.putDouble(STATE_BAVG_QUAL, bavg_quality);
        savedInstanceState.putDouble(STATE_BILL, bill);
        savedInstanceState.putDouble(STATE_SERVICE_QUAL, service_quality);

        savedInstanceState.putInt(STATE_NUM_PEOPLE, num_people);
        savedInstanceState.putBoolean(STATE_RADIO_CHECK, radio_checked);


        super.onSaveInstanceState(savedInstanceState);
    }
    ActivityResultLauncher<Intent>
            someActivityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            new ActivityResultCallback<ActivityResult>() {
                @Override
                public void onActivityResult(ActivityResult result)
                {
                    if (result.getResultCode() == Activity.RESULT_OK) {
                        Intent data = result.getData();
                        exc_quality = data.getExtras().getDouble(RESULT1);
                        avg_quality = data.getExtras().getDouble(RESULT2);
                        bavg_quality = data.getExtras().getDouble(RESULT3);
                        Toast.makeText(getApplicationContext(), String.format("%f %f %f", exc_quality, avg_quality, bavg_quality), Toast.LENGTH_LONG).show();
                        updateServiceQuality();
                    }
                }
            });

}

File: lab7pbudhath/app/src/main/java/com/example/lab7_pbudhath/MainActivity.kt
package com.example.lab7_pbudhath

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.RadioButton
import android.widget.TextView
import android.widget.Toast

class MainActivity : AppCompatActivity() {

    private var bill: Float = 0.0f
    private var serviceQuality: Float = 0.0f
    private var excQuality: Float = 0.25f
    private var avgQuality: Float = 0.20f
    private var bavgQuality: Float = 0.15f
    private var numPeople: Int = 1
    private var radioChecked: Boolean = false

    companion object {
        const val STATE_EXC_QUAL: String = "EXC_QUALITY"
        const val STATE_AVG_QUAL: String = "AVG_QUALITY"
        const val STATE_BAVG_QUAL: String = "BAVG_QUALITY"
        const val STATE_RADIO_CHECK: String = "RADIO_CHECKED"
        const val STATE_NUM_PEOPLE: String = "NUM_PEOPLE"
        const val STATE_SERVICE_QUAL: String = "SERVICE_QUALITY"
        const val STATE_BILL: String = "BILL"

        const val RESULT1: String = "RESULT1"
        const val RESULT2: String = "RESULT2"
        const val RESULT3: String = "RESULT3"

    }
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if(savedInstanceState != null){
            excQuality = savedInstanceState.getFloat(STATE_EXC_QUAL)
            avgQuality = savedInstanceState.getFloat(STATE_AVG_QUAL)
            bavgQuality = savedInstanceState.getFloat(STATE_BAVG_QUAL)

            serviceQuality = savedInstanceState.getFloat(STATE_SERVICE_QUAL)
            bill = savedInstanceState.getFloat(STATE_SERVICE_QUAL)
            numPeople = savedInstanceState.getInt(STATE_SERVICE_QUAL)
            radioChecked = savedInstanceState.getBoolean(STATE_SERVICE_QUAL)

        }

        val billEdit = findViewById<EditText>(R.id.Bill_Edit)
        val peopleEdit = findViewById<EditText>(R.id.People_Edit)
        val twb = object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                if (billEdit.text.isNotEmpty()){
                    bill = billEdit.text.toString().toFloat()
                    if(bill > 10000){
                        Toast.makeText(applicationContext, "Max is $10,000",
                            Toast.LENGTH_SHORT).show()
                        bill = 0f
                        return
                    }

                    if(bill < 0){
                        Toast.makeText(applicationContext, "Invalid Entry",
                            Toast.LENGTH_SHORT).show()
                        bill = 0f
                        return
                    }
                    if(radioChecked){
                        updateText()
                    }
                    else{
                        Toast.makeText(applicationContext, "Select a service quality",
                            Toast.LENGTH_SHORT).show()
                    }
                }
                else{
                    bill = 0f
                    findViewById<TextView>(R.id.Tip_Total_Val).text = ""
                    findViewById<TextView>(R.id.Total_Val).text = ""
                    findViewById<TextView>(R.id.Total_Per_Val).text = ""
                }
            }
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        }
        billEdit.addTextChangedListener(twb)
        billEdit.setText("")

        val twp = object : TextWatcher {
            override fun afterTextChanged(s: Editable?) {
                if (peopleEdit.text.isNotEmpty()){
                    numPeople = peopleEdit.text.toString().toInt()
                    if(numPeople > 10){
                        Toast.makeText(applicationContext, "Max is $10,000",
                            Toast.LENGTH_SHORT).show()
                        numPeople = 1
                        return
                    }

                    if(numPeople < 1){
                        Toast.makeText(applicationContext, "Invalid Entry",
                            Toast.LENGTH_SHORT).show()
                        numPeople = 1
                        return
                    }
                    if(radioChecked){
                        updateText()
                    }
                    else{
                        Toast.makeText(applicationContext, "Select a service quality",
                            Toast.LENGTH_SHORT).show()
                    }
                }
                else{
                    numPeople = 1
                    findViewById<TextView>(R.id.Tip_Total_Val).text = ""
                    findViewById<TextView>(R.id.Total_Val).text = ""
                    findViewById<TextView>(R.id.Total_Per_Val).text = ""
                }
            }
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
        }

        peopleEdit.addTextChangedListener(twp)
        peopleEdit.setText("")

    }

    fun myOnClick(v: View){
        if(v.id == R.id.Clear_Button){
            findViewById<EditText>(R.id.Bill_Edit).setText("")
            findViewById<EditText>(R.id.People_Edit).setText("")
            findViewById<RadioButton>(R.id.Radio_Exc).isChecked = false
            findViewById<RadioButton>(R.id.Radio_Avg).isChecked = false
            findViewById<RadioButton>(R.id.Radio_Bavg).isChecked = false
            findViewById<TextView>(R.id.Tip_Total_Val).text = ""
            findViewById<TextView>(R.id.Total_Val).text = ""
            findViewById<TextView>(R.id.Total_Per_Val).text = ""
            radioChecked = false
            bill = 0.0f
        }

        if(v.id == R.id.Update_Button){
            val intent = Intent(applicationContext, ServiceSelectActivity::class.java)
            intent.putExtra(STATE_EXC_QUAL, excQuality)
            intent.putExtra(STATE_AVG_QUAL, avgQuality)
            intent.putExtra(STATE_BAVG_QUAL, bavgQuality)
            startActivityForResult(intent, 0)
        }
    }

    fun onServiceRadioClick(v: View){
        radioChecked = (v as RadioButton).isChecked
        if(!radioChecked){
            serviceQuality = 0f
            return;
        }

        if(v.id == R.id.Radio_Exc){
            serviceQuality = excQuality
        }
        if(v.id == R.id.Radio_Avg){
            serviceQuality = avgQuality
        }
        if(v.id == R.id.Radio_Bavg){
            serviceQuality = bavgQuality
        }

        if(bill == 0f){
            Toast.makeText(applicationContext, "Enter bill amount", Toast.LENGTH_SHORT).show();
        }
        else{
            updateText()
        }
    }

    fun updateText(){

        if(!radioChecked){
            return;
        }

        if(bill == 0f){
            return;
        }

        val tip_total: Float =  bill * serviceQuality;
        val total: Float = bill + tip_total;
        val total_per: Float = total/numPeople;

        (findViewById<TextView>(R.id.Tip_Total_Val)).text = (String.format("%.2f%n", tip_total))
        (findViewById<TextView>(R.id.Total_Val)).text = (String.format("%.2f%n", total))
        (findViewById<TextView>(R.id.Total_Per_Val)).text = (String.format("%.2f%n", total_per))
    }

    private fun updateServiceQuality(){
        (findViewById<TextView>(R.id.Radio_Exc))
            .text = (String.format("Excellent (%.2f%%)", excQuality*100))
        (findViewById<TextView>(R.id.Radio_Avg))
            .text = (String.format("Average (%.2f%%)", avgQuality*100))
        (findViewById<TextView>(R.id.Radio_Bavg))
            .text =(String.format("Below Average (%.2f%%)", bavgQuality*100))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?){
        super.onActivityResult(requestCode, resultCode, data)
        if(data != null) {
            excQuality = data.getFloatExtra(RESULT1, 0.25f)
            avgQuality = data.getFloatExtra(RESULT2, 0.20f)
            bavgQuality = data.getFloatExtra(RESULT3, 0.15f)
            updateServiceQuality()
        }

    }
    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        savedInstanceState.putFloat(STATE_EXC_QUAL, excQuality);
        savedInstanceState.putFloat(STATE_AVG_QUAL, avgQuality);
        savedInstanceState.putFloat(STATE_BAVG_QUAL, bavgQuality);
        savedInstanceState.putFloat(STATE_BILL, bill);
        savedInstanceState.putFloat(STATE_SERVICE_QUAL, serviceQuality);

        savedInstanceState.putInt(STATE_NUM_PEOPLE, numPeople);
        savedInstanceState.putBoolean(STATE_RADIO_CHECK, radioChecked);

        super.onSaveInstanceState(savedInstanceState);
    }
}